From a92099b60b60d201f6a419de0a61968967718d1e Mon Sep 17 00:00:00 2001
From: Ayoub Zaki <ayoub.zaki@embetrix.com>
Date: Wed, 31 Jul 2024 17:49:33 +0200
Subject: [PATCH 2/4] shift kernel load address

Signed-off-by: Ayoub Zaki <ayoub.zaki@embetrix.com>
---
 include/configs/stm32mp15_common.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/configs/stm32mp15_common.h b/include/configs/stm32mp15_common.h
index 50a989332d..55ad90c54d 100644
--- a/include/configs/stm32mp15_common.h
+++ b/include/configs/stm32mp15_common.h
@@ -114,7 +114,7 @@
  * 1M fdt, 1M script, 1M pxe and 1M for overlay
  * and the ramdisk at the end.
  */
-#define __KERNEL_ADDR_R     __stringify(0xc2000000)
+#define __KERNEL_ADDR_R     __stringify(0xc8000000)
 #define __FDT_ADDR_R        __stringify(0xc6000000)
 #define __SCRIPT_ADDR_R     __stringify(0xc6100000)
 #define __PXEFILE_ADDR_R    __stringify(0xc6200000)
-- 
2.34.1

