diff --git a/configs/stm32mp15_defconfig b/configs/stm32mp15_defconfig
index be25aaa788..54589a7163 100644
--- a/configs/stm32mp15_defconfig
+++ b/configs/stm32mp15_defconfig
@@ -2,14 +2,14 @@ CONFIG_ARM=y
 CONFIG_ARCH_STM32MP=y
 CONFIG_TFABOOT=y
 CONFIG_SYS_MALLOC_F_LEN=0x80000
-CONFIG_ENV_OFFSET=0x900000
+CONFIG_ENV_OFFSET=0x280000
 CONFIG_ENV_SECT_SIZE=0x40000
 CONFIG_DEFAULT_DEVICE_TREE="stm32mp157c-ev1"
 CONFIG_SYS_PROMPT="STM32MP> "
 CONFIG_DDR_CACHEABLE_SIZE=0x8000000
 CONFIG_CMD_STM32KEY=y
 CONFIG_TARGET_ST_STM32MP15X=y
-CONFIG_ENV_OFFSET_REDUND=0x940000
+CONFIG_ENV_OFFSET_REDUND=0x2C0000
 CONFIG_CMD_STM32PROG=y
 # CONFIG_ARMV7_NONSEC is not set
 CONFIG_SYS_LOAD_ADDR=0xc2000000
@@ -63,7 +63,7 @@ CONFIG_ENV_UBI_PART="UBI"
 CONFIG_ENV_UBI_VOLUME="uboot_config"
 CONFIG_ENV_UBI_VOLUME_REDUND="uboot_config_r"
 CONFIG_SYS_RELOC_GD_ENV_ADDR=y
-CONFIG_SYS_MMC_ENV_DEV=-1
+CONFIG_SYS_MMC_ENV_DEV=0
 CONFIG_ENV_MMC_USE_DT=y
 CONFIG_TFTP_TSIZE=y
 CONFIG_STM32_ADC=y
@@ -171,3 +171,10 @@ CONFIG_ERRNO_STR=y
 # CONFIG_LMB_USE_MAX_REGIONS is not set
 CONFIG_LMB_MEMORY_REGIONS=2
 CONFIG_LMB_RESERVED_REGIONS=16
+CONFIG_CMD_BOOTCOUNT=y
+CONFIG_BOOTCOUNT_LIMIT=y
+CONFIG_BOOTCOUNT_GENERIC=y
+CONFIG_SYS_BOOTCOUNT_MAGIC=0xB001C041
+CONFIG_BOOTCOUNT_BOOTLIMIT=3
+CONFIG_SYS_BOOTCOUNT_ADDR=0x5C00A154
+CONFIG_SYS_BOOTCOUNT_SINGLEWORD=y
diff --git a/include/configs/stm32mp15_common.h b/include/configs/stm32mp15_common.h
index bd8e16bc1b..ecd00327c2 100644
--- a/include/configs/stm32mp15_common.h
+++ b/include/configs/stm32mp15_common.h
@@ -123,7 +123,7 @@
  * 1M fdt, 1M script, 1M pxe and 1M for overlay
  * and the ramdisk at the end.
  */
-#define __KERNEL_ADDR_R     __stringify(0xc2000000)
+#define __KERNEL_ADDR_R     __stringify(0xc8000000)
 #define __FDT_ADDR_R        __stringify(0xc4000000)
 #define __SCRIPT_ADDR_R     __stringify(0xc4100000)
 #define __PXEFILE_ADDR_R    __stringify(0xc4200000)
diff --git a/include/configs/stm32mp15_st_common.h b/include/configs/stm32mp15_st_common.h
index e9bfbc482e..df9a9387b8 100644
--- a/include/configs/stm32mp15_st_common.h
+++ b/include/configs/stm32mp15_st_common.h
@@ -28,21 +28,24 @@
  * for nor boot, distro boot on SD card = mmc0 ONLY !
  */
 #define ST_STM32MP1_BOOTCMD "bootcmd_stm32mp=" \
-	"echo \"Boot over ${boot_device}${boot_instance}!\";" \
-	"if test ${boot_device} = serial || test ${boot_device} = usb;" \
-	"then stm32prog ${boot_device} ${boot_instance}; " \
+	"if test ${boot_device} = serial || test ${boot_device} = usb; then " \
+		"stm32prog ${boot_device} ${boot_instance}; " \
 	"else " \
-		"run env_check;" \
-		"if test ${boot_device} = mmc;" \
-		"then env set boot_targets \"mmc${boot_instance}\"; fi;" \
-		"if test ${boot_device} = nand ||" \
-		  " test ${boot_device} = spi-nand ;" \
-		"then env set boot_targets ubifs0; fi;" \
-		"if test ${boot_device} = nor;" \
-		"then env set boot_targets mmc0; fi;" \
-		"run distro_bootcmd;" \
+		"if test ${boot_device} = mmc; then run sd_boot; fi;" \
 	"fi;\0"
 
+#define SD_BOOT \
+	"sd_boot=if test ${active_system} = 1; then setenv mmcroot /dev/mmcblk0p5 && setenv bootpart 0:5; fi;" \
+	"ext4load mmc ${bootpart} ${kernel_addr_r} /boot/${image.linux};" \
+	"setenv bootargs \"console=${console} root=${mmcroot} ${optargs}\";" \
+	"bootm ${kernel_addr_r}\0"
+
+#define NET_BOOT \
+	"net_boot=dhcp; setenv serverip ${gatewayip};" \
+	"tftp ${kernel_addr_r} ${image.linux};" \
+	"setenv bootargs \"console=${console} root=${mmcroot} ${optargs}\";" \
+	"bootm ${kernel_addr_r}\0"
+
 #undef CONFIG_EXTRA_ENV_SETTINGS
 #define CONFIG_EXTRA_ENV_SETTINGS \
 	STM32MP_MEM_LAYOUT \
@@ -50,7 +53,18 @@
 	STM32MP_PARTS_DEFAULT \
 	BOOTENV \
 	STM32MP_EXTRA \
-	STM32MP_BOARD_EXTRA_ENV
+	STM32MP_BOARD_EXTRA_ENV \
+	"bootdelay=1\0" \
+	"bootpart=0:4\0" \
+	"active_system=0\0" \
+	"console=ttySTM0,115200\0" \
+	"image.linux=fitImage\0" \
+	"optargs=rw rootwait loglevel=7 vt.global_cursor_default=0 fbcon=rotate:3\0" \
+	"mmcroot=/dev/mmcblk0p4\0" \
+	"bootlimit=3\0" \
+	"altbootcmd=if test ${active_system} = 1; then setenv active_system 0; else setenv active_system 1; fi; saveenv; run bootcmd\0" \
+	SD_BOOT \
+	NET_BOOT
 
 #endif
 #endif
