diff --git a/configs/stm32mp15_defconfig b/configs/stm32mp15_defconfig
index b4b2954273..ac8195b245 100644
--- a/configs/stm32mp15_defconfig
+++ b/configs/stm32mp15_defconfig
@@ -4,13 +4,13 @@ CONFIG_TFABOOT=y
 CONFIG_SYS_MALLOC_F_LEN=0x80000
 CONFIG_SYS_MEMTEST_START=0xc0000000
 CONFIG_SYS_MEMTEST_END=0xc4000000
-CONFIG_ENV_OFFSET=0x900000
+CONFIG_ENV_OFFSET=0x280000
 CONFIG_ENV_SECT_SIZE=0x40000
 CONFIG_DEFAULT_DEVICE_TREE="stm32mp157c-ev1"
 CONFIG_DDR_CACHEABLE_SIZE=0x10000000
 CONFIG_CMD_STM32KEY=y
 CONFIG_TARGET_ST_STM32MP15x=y
-CONFIG_ENV_OFFSET_REDUND=0x940000
+CONFIG_ENV_OFFSET_REDUND=0x2C0000
 CONFIG_CMD_STM32PROG=y
 # CONFIG_ARMV7_NONSEC is not set
 CONFIG_DISTRO_DEFAULTS=y
@@ -162,3 +162,10 @@ CONFIG_FDT_FIXUP_PARTITIONS=y
 # CONFIG_LMB_USE_MAX_REGIONS is not set
 CONFIG_LMB_MEMORY_REGIONS=2
 CONFIG_LMB_RESERVED_REGIONS=16
+CONFIG_CMD_BOOTCOUNT=y
+CONFIG_BOOTCOUNT_LIMIT=y
+CONFIG_BOOTCOUNT_GENERIC=y
+CONFIG_SYS_BOOTCOUNT_MAGIC=0xB001C041
+CONFIG_BOOTCOUNT_BOOTLIMIT=3
+CONFIG_SYS_BOOTCOUNT_ADDR=0x5C00A154
+CONFIG_SYS_BOOTCOUNT_SINGLEWORD=y
diff --git a/include/configs/stm32mp15_common.h b/include/configs/stm32mp15_common.h
index 58971cc656..df36d19f94 100644
--- a/include/configs/stm32mp15_common.h
+++ b/include/configs/stm32mp15_common.h
@@ -159,7 +159,7 @@
  * 1M fdt, 1M script, 1M pxe and 1M for overlay
  * and the ramdisk at the end.
  */
-#define __KERNEL_ADDR_R     __stringify(0xc2000000)
+#define __KERNEL_ADDR_R     __stringify(0xc8000000)
 #define __FDT_ADDR_R        __stringify(0xc4000000)
 #define __SCRIPT_ADDR_R     __stringify(0xc4100000)
 #define __PXEFILE_ADDR_R    __stringify(0xc4200000)
diff --git a/include/configs/stm32mp15_st_common.h b/include/configs/stm32mp15_st_common.h
index c395f7f986..a1de1bc8e6 100644
--- a/include/configs/stm32mp15_st_common.h
+++ b/include/configs/stm32mp15_st_common.h
@@ -27,22 +27,25 @@
  * for nand or spi-nand boot, distro boot with ubifs on UBI partition
  * for nor boot, distro boot on SD card = mmc0 ONLY !
  */
-#define ST_STM32MP1_BOOTCMD "bootcmd_stm32mp=" \
-	"echo \"Boot over ${boot_device}${boot_instance}!\";" \
-	"if test ${boot_device} = serial || test ${boot_device} = usb;" \
-	"then stm32prog ${boot_device} ${boot_instance}; " \
+#define ST_STM32MP1_BOOTCMD "bootcmd_stm32mp=" \
+	"if test ${boot_device} = serial || test ${boot_device} = usb; then " \
+		"stm32prog ${boot_device} ${boot_instance}; " \
 	"else " \
-		"run env_check;" \
-		"if test ${boot_device} = mmc;" \
-		"then env set boot_targets \"mmc${boot_instance}\"; fi;" \
-		"if test ${boot_device} = nand ||" \
-		  " test ${boot_device} = spi-nand ;" \
-		"then env set boot_targets ubifs0; fi;" \
-		"if test ${boot_device} = nor;" \
-		"then env set boot_targets mmc0; fi;" \
-		"run distro_bootcmd;" \
+		"if test ${boot_device} = mmc; then run sd_boot; fi;" \
 	"fi;\0"
 
+#define SD_BOOT \
+	"sd_boot=if test ${active_system} = 1; then setenv mmcroot /dev/mmcblk0p5 && setenv bootpart 0:5; fi;" \
+	"ext4load mmc ${bootpart} ${kernel_addr_r} /boot/${image.linux};" \
+	"setenv bootargs \"console=${console} root=${mmcroot} ${optargs}\";" \
+	"bootm ${kernel_addr_r}\0"
+
+#define NET_BOOT \
+	"net_boot=dhcp; setenv serverip ${gatewayip};" \
+	"tftp ${kernel_addr_r} ${image.linux};" \
+	"setenv bootargs \"console=${console} root=${mmcroot} ${optargs}\";" \
+	"bootm ${kernel_addr_r}\0"
+
 #undef CONFIG_EXTRA_ENV_SETTINGS
 #define CONFIG_EXTRA_ENV_SETTINGS \
 	STM32MP_MEM_LAYOUT \
@@ -50,7 +53,18 @@
 	STM32MP_PARTS_DEFAULT \
 	BOOTENV \
 	STM32MP_EXTRA \
-	STM32MP_BOARD_EXTRA_ENV
+	STM32MP_BOARD_EXTRA_ENV \
+	"bootdelay=1\0" \
+	"bootpart=0:4\0" \
+	"active_system=0\0" \
+	"console=ttySTM0,115200\0" \
+	"image.linux=fitImage\0" \
+	"optargs=rw rootwait quiet loglevel=0 vt.global_cursor_default=0 fbcon=rotate:3\0" \
+	"mmcroot=/dev/mmcblk0p4\0" \
+	"bootlimit=3\0" \
+	"altbootcmd=if test ${active_system} = 1; then setenv active_system 0; else setenv active_system 1; fi; saveenv; run bootcmd\0" \
+	SD_BOOT \
+	NET_BOOT
 
 #endif
 #endif
